device/thunder_ripple_sdk/src/client/thunder_async_client.rs

```diff
@@ -0,0 +1,304 @@
+use futures_util::{SinkExt, StreamExt};
+use ripple_sdk::{
+    api::{
+        device::device_operator::DeviceChannelRequest, gateway::rpc_gateway_api::JsonRpcApiResponse,
+    },
+    log::{debug, error, info},
+    tokio::{
+        self,
+        net::TcpStream,
+        sync::mpsc::{Receiver, Sender},
+    },
+    utils::{error::RippleError, rpc_utils::extract_tcp_port},
+};
+
+use super::thunder_plugins_status_mgr::{BrokerCallback, BrokerSender, StatusManager};
+use crate::utils::get_next_id;
+use serde_json::json;
+use tokio_tungstenite::{client_async, tungstenite::Message, WebSocketStream};
+
+#[derive(Clone, Debug)]
+pub struct ThunderAsyncClient {
+    status_manager: StatusManager,
+    sender: BrokerSender,
+    callback: BrokerCallback,
+}
+
+#[derive(Clone, Debug)]
+pub struct ThunderAsyncRequest {
+    pub id: u64,
+    request: DeviceChannelRequest,
+}
+
+impl ThunderAsyncRequest {
+    pub fn new(request: DeviceChannelRequest) -> Self {
+        Self {
+            id: get_next_id(),
+            request,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ThunderAsyncResponse {
+    pub id: Option<u64>,
+    pub result: Result<JsonRpcApiResponse, RippleError>,
+}
+
+impl ThunderAsyncResponse {
+    fn new_response(response: JsonRpcApiResponse) -> Self {
+        Self {
+            id: None,
+            result: Ok(response),
+        }
+    }
+
+    fn new_error(id: u64, e: RippleError) -> Self {
+        Self {
+            id: Some(id),
+            result: Err(e),
+        }
+    }
+}
+
+impl ThunderAsyncClient {
+    pub fn get_sender(&self) -> BrokerSender {
+        self.sender.clone()
+    }
+
+    async fn create_ws(
+        endpoint: &str,
+    ) -> (
+        SplitSink<WebSocketStream<TcpStream>, Message>,
+        SplitStream<WebSocketStream<TcpStream>>,
+    ) {
+        info!("Broker Endpoint url {}", endpoint);
+
+        let url = url::Url::parse(endpoint).unwrap();
+        let port = extract_tcp_port(endpoint);
+        info!("Url host str {}", url.host_str().unwrap());
+        let mut index = 0;
+
+        loop {
+            // Try connecting to the tcp port first
+            if let Ok(v) = TcpStream::connect(&port).await {
+                // Setup handshake for websocket with the tcp port
+                // Some WS servers lock on to the Port but not setup handshake till they are fully setup
+                if let Ok((stream, _)) = client_async(endpoint, v).await {
+                    break stream.split();
+                }
+            }
+            if (index % 10).eq(&0) {
+                error!(
+                    "Broker with {} failed with retry for last {} secs in {}",
+                    endpoint, index, port
+                );
+            }
+            index += 1;
+            tokio::time::sleep(Duration::from_secs(1)).await;
+        }
+    }
+
+    fn prepare_request(&self, request: &ThunderAsyncRequest) -> Result<Vec<String>, RippleError> {
```

Comment:
Do you have any documentation related to the new changes ?  Is ThunderBroker a singleton object ?  If not, I have concern about multiple instances of ThunderBroker.  In that case you may need to separate state management instance out of Thunder Broker. 

I have also cleaned up  `prepare_request` in the original implementation.  Thunder Plugin state checking and activation have been moved to `check_and_generate_plugin_activation_request`  helper function,